session/session.go

```diff
@@ -346,96 +295,143 @@ func (s *Session) handlePeriodicSearch(ctx context.Context) {
+	return live
 }
 
-func (s *Session) cidIsWanted(c cid.Cid) bool {
-	_, ok := s.liveWants[c]
+func (s *Session) unlockedIsWanted(c cid.Cid) bool {
+	_, ok := s.sw.liveWants[c]
 	if !ok {
-		ok = s.tofetch.Has(c)
+		ok = s.sw.toFetch.Has(c)
 	}
 	return ok
 }
 
-func (s *Session) processIncoming(ctx context.Context, ks []cid.Cid) {
-	for _, c := range ks {
-		if s.cidIsWanted(c) {
-			// If the block CID was in the live wants queue, remove it
-			tval, ok := s.liveWants[c]
-			if ok {
-				s.latTotal += time.Since(tval)
-				delete(s.liveWants, c)
-			} else {
-				// Otherwise remove it from the tofetch queue, if it was there
-				s.tofetch.Remove(c)
-			}
-			s.fetchcnt++
-
-			// We've received new wanted blocks, so reset the number of ticks
-			// that have occurred since the last new block
-			s.consecutiveTicks = 0
-
-			// Keep track of CIDs we've successfully fetched
-			s.pastWants.Push(c)
-		}
+func (s *Session) handleIncoming(ctx context.Context, rcv rcvFrom) {
+	// Record statistics only if the blocks came from the network
+	// (blocks can also be received from the local node)
+	if rcv.from != "" {
+		s.updateReceiveCounters(ctx, rcv)
 	}
 
-	// Transfer as many CIDs as possible from the tofetch queue into the
-	// live wants queue
-	toAdd := s.wantBudget()
-	if toAdd > s.tofetch.Len() {
-		toAdd = s.tofetch.Len()
-	}
-	if toAdd > 0 {
-		var keys []cid.Cid
-		for i := 0; i < toAdd; i++ {
-			keys = append(keys, s.tofetch.Pop())
-		}
-		s.wantBlocks(ctx, keys)
+	// Update the want list
+	wanted, totalLatency := s.blocksReceived(rcv.ks)
+	if len(wanted) == 0 {
+		return
 	}
+
+	// We've received the blocks so we can cancel any outstanding wants for them
+	s.cancelIncoming(ctx, wanted)
+
+	s.idleTick.Stop()
+
+	// Process the received blocks
+	s.processIncoming(ctx, wanted, totalLatency)
+
+	s.resetIdleTick()
 }
 
 func (s *Session) updateReceiveCounters(ctx context.Context, rcv rcvFrom) {
-	for _, k := range rcv.ks {
-		// Inform the request splitter of unique / duplicate blocks
-		if s.cidIsWanted(k) {
+	s.sw.RLock()
+
+	for _, c := range rcv.ks {
+		if s.unlockedIsWanted(c) {
 			s.srs.RecordUniqueBlock()
-		} else if s.pastWants.Has(k) {
+		} else if s.sw.pastWants.Has(c) {
 			s.srs.RecordDuplicateBlock()
 		}
 	}
 
+	s.sw.RUnlock()
+
 	// Record response (to be able to time latency)
 	if len(rcv.ks) > 0 {
 		s.pm.RecordPeerResponse(rcv.from, rcv.ks)
 	}
 }
 
-func (s *Session) wantBlocks(ctx context.Context, ks []cid.Cid) {
-	now := time.Now()
-	for _, c := range ks {
-		s.liveWants[c] = now
+func (s *Session) blocksReceived(cids []cid.Cid) ([]cid.Cid, time.Duration) {
+	s.sw.Lock()
+	defer s.sw.Unlock()
+
+	totalLatency := time.Duration(0)
+	wanted := make([]cid.Cid, 0, len(cids))
+	for _, c := range cids {
+		if s.unlockedIsWanted(c) {
+			wanted = append(wanted, c)
+
+			// If the block CID was in the live wants queue, remove it
+			tval, ok := s.sw.liveWants[c]
+			if ok {
+				totalLatency += time.Since(tval)
```

Comment:
Let's avoid taking the time multiple times inside the lock (makes a syscall on some platforms). We can just take it once outside.